### PR TITLE
drivers: pinctrl: Update Atmel sam/sam0 pinctrl drivers

### PR DESCRIPTION
**Pull request by nandojve**
This series implement Atmel sam/sam0 pinctrl. It update old version in favor of new API and groups devicetree. In addition it drops sam0 pinmux driver and sam boards pinmux. All drivers that somehow may use pins are converted to use pinctrl and not necessarily have entries right now at boards devicetree files.

This is Atmel sam/sam0 part of #39740.

Tested with:
 - atsamd20_xpro
 - atsamr21_xpro
 - sam4s_xplained
 - sam4e_xpro
 - sam4l_ek
 - sam_v71_xult

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]
LGTM, excellent work!

**Comment by nandojve** [comment]
rebase due to conflicts.

**Review by gmarull — DISMISSED** [review]

File: drivers/pinctrl/pinctrl_sam.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/pinctrl.h>
+#include <soc_gpio.h>
+
+/** Utility macro that expands to the GPIO port address if it exists */
+#define SAM_PORT_ADDR_OR_NONE(nodelabel)					\
+	COND_CODE_1(DT_NODE_EXISTS(DT_NODELABEL(nodelabel)),			\
+		   (DT_REG_ADDR(DT_NODELABEL(nodelabel)),), ())
```
Comment (mbolivar-nordic) [inline drivers/pinctrl/pinctrl_sam.c:13]: Just FYI, you can use `IF_ENABLED(DT_NODE_EXISTS(DT_NODELABEL(nodelabel)), (DT_REG_ADDR(...)))` instead if you want to save some typing.

File: dts/bindings/pinctrl/atmel,sam-pinctrl.yaml [hunk 2]
```diff
@@ -1,22 +1,85 @@
...
+          /* configure PA6 as USART0 TX and PA8 as USART0 CTS */
+          pinmux = <PA6A_USART0_TXD0>, <PA8A_USART0_CTS0>;
+        };
+        /* group 2 */
+        group2 {
+          /* configure PA5 as USART0 RX and PA7 as USART0 RTS */
+          pinmux = <PA5A_USART0_RXD0>, <PA7A_USART0_RTS0>;
+          /* both PA5 and PA7 have pull-up enabled */
+          bias-pull-up;
+        };
```
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/atmel,sam-pinctrl.yaml:44]: Missing `}; };` at the end here, right?

File: dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml [hunk 3]
```diff
@@ -1,34 +1,89 @@
...
+          /* configure PA6 as USART0 TX and PA8 as USART0 CTS */
+          pinmux = <PA5D_SERCOM0_PAD1>, <PA6D_SERCOM0_PAD2>;
+        };
+        /* group 2 */
+        group2 {
+          /* configure PA5 as USART0 RX and PA7 as USART0 RTS */
+          pinmux = <PA4D_SERCOM0_PAD0>, <PA7D_SERCOM0_PAD3>;
+          /* both PA5 and PA7 have pull-up enabled */
+          bias-pull-up;
+        };
```
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml:44]: Missing `}; };`.

**Review by mbolivar-nordic — COMMENTED** [review]
Great work. I noticed some minor typos that will affect users of the pinctrl  binding in the YAML descriptions. I would like to see those fixed before merge, but otherwise I'm +1.

File: drivers/pinctrl/pinctrl_sam.c (hunk 1 above)
Comment (nandojve) [reply drivers/pinctrl/pinctrl_sam.c]: Yeah, it makes cleaner, tks

File: dts/bindings/pinctrl/atmel,sam-pinctrl.yaml (hunk 2 above)
Comment (nandojve) [reply dts/bindings/pinctrl/atmel,sam-pinctrl.yaml:44]: Ops, definitely!

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by gmarull** [comment]
@nandojve added a comment about spdx in the HAL PR (`Apache 2.0` -> `Apache-2.0`), please check.

**Review by gmarull — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**